### ISSUE TITLE
Source generated endpoint can not process messages sent from send-only endpoint

### ISSUE THREAD

**Issue by simonvandermeer** [Particular/NServiceBus.AzureFunctions.Worker.ServiceBus#418]
#### Description
A message sent from a send-only endpoint to a function endpoint which is generated using the NServiceBusTriggerFunction attribute can not be processed.

#### Expected behavior
Messages from send-only endpoints can be processed.

#### Actual behavior
Exception is thrown.

#### Versions
Microsoft.Azure.Functions.Worker: 1.20.0
NServiceBus.AzureFunctions.Worker.ServiceBus: 4.2.1

### Steps to reproduce

Attempt to process a message from a send-only endpoint on a function endpoint generated with the NServiceBusTriggerFunction attribute.

### Relevant log output

```shell
Function 'NServiceBusTrigger', Invocation id '961c61af-7649-4d1d-87ae-c39c7427235a': An exception was thrown by the invocation.
[2023-11-28T13:09:30.159Z] Result: Function 'NServiceBusTrigger', Invocation id '961c61af-7649-4d1d-87ae-c39c7427235a': An exception was thrown by the invocation.
Exception: Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'NServiceBusTrigger': Could not populate the value for 'replyTo' parameter. Consider updating the parameter with a default value.
[2023-11-28T13:09:30.160Z]    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInputBindingFeature.cs:line 88
[2023-11-28T13:09:30.160Z]    at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 42
[2023-11-28T13:09:30.161Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBindings\OutputBindingsMiddleware.cs:line 13
[2023-11-28T13:09:30.159Z] Executed 'Functions.NServiceBusTrigger' (Failed, Id=961c61af-7649-4d1d-87ae-c39c7427235a, Duration=4ms)
[2023-11-28T13:09:30.161Z]    at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 77
Stack:    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInputBindingFeature.cs:line 88
[2023-11-28T13:09:30.163Z]    at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 42
[2023-11-28T13:09:30.163Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBindings\OutputBindingsMiddleware.cs:line 13
[2023-11-28T13:09:30.162Z] System.Private.CoreLib: Exception while executing function: Functions.NServiceBusTrigger. System.Private.CoreLib: Result: Failure
Exception: Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'NServiceBusTrigger': Could not populate the value for 'replyTo' parameter. Consider updating the parameter with a default value.
[2023-11-28T13:09:30.165Z]    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInputBindingFeature.cs:line 88
[2023-11-28T13:09:30.165Z]    at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 42
[2023-11-28T13:09:30.164Z] Executed 'Functions.NServiceBusTrigger' (Failed, Id=a74c3a06-9584-40e1-bc30-d1b71aeac24f, Duration=0ms)
[2023-11-28T13:09:30.164Z]    at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 77.
```

### Additional Information

[This regression in Microsoft.Azure.Functions.Worker >= 1.20](https://github.com/Azure/azure-functions-dotnet-worker/issues/2065) seems to be the culprit. Once I downgrade the package to 1.19.0 the exception no longer occures and messages are processed.

Another workaround is manually creating the service bus trigger and setting a null default on the replyTo argument. This workaround is also mentioned in the referenced issue.
```csharp
public class FuseEndpointTrigger
{
    private readonly IFunctionEndpoint _endpoint;

    public FuseEndpointTrigger(IFunctionEndpoint endpoint)
    {
        _endpoint = endpoint;
    }

    [Function(nameof(FuseEndpointTrigger))]
    public async Task Run(
        [ServiceBusTrigger("fdr.wm.wom")] byte[] messageBody,
        IDictionary<string, object> userProperties,
        FunctionContext context,
        string messageId,
        int deliveryCount,
        string correlationId,
        string replyTo = null)
    {
        await _endpoint.Process(messageBody, userProperties, messageId, deliveryCount, replyTo, correlationId, context);
    }
}
```

**Comment by danielmarbach** [comment]
Thank you @simonvandermeer for reporting the bug with so many details. Indeed, there appears to be a regression in the Microsoft.Azure.Functions.Worker package. We have already seen other ramifications of the potential regression in 
https://github.com/Particular/NServiceBus.AzureFunctions.Worker.ServiceBus/issues/415

I'm still trying to build up a more in-depth understanding of the underlying root cause. At first sight, it seems like the change is that the binder no longer provides a default value when a [MessageMetadata](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-service-bus-trigger?tabs=python-v2%2Cisolated-process%2Cnodejs-v4%2Cextensionv5&pivots=programming-language-csharp#message-metadata) property is not set which apparently it has been doing before the regression. ReplyAddress is only set by NServiceBus in a regular endpoint. CorrelationId is always set even for send-only endpoints.

In case Microsoft will not fix the behavior we are seeing, we might need to update the code to explicitly provide a default value by changing `string replyTo = null` which aligns with the workaround mentioned in https://github.com/Azure/azure-functions-dotnet-worker/issues/2065#issuecomment-1825225609.

**Comment by danielmarbach** [comment]
We have merged a fix for the send only problem to master and current supported release branches. But there seems to be still a fundamental problem with the combination of the SDK and the worker packages that we need to investigate further before we release.

**Comment by danielmarbach** [comment]
Heads up, the issue seems to be related to the new metadata provider source generator and the order in which source generators get invoked see https://github.com/Particular/NServiceBus.AzureFunctions.Worker.ServiceBus/issues/428#issuecomment-1875600156

**Comment by danielmarbach** [comment]
[Microsoft has released 1.20.1 which addresses this problem too](https://github.com/Azure/azure-functions-dotnet-worker/issues/2065#issuecomment-1887773592). We will still proceed with the changes we added, but we are processing some more adjustments to make the source generator work again.
